### ISSUE TITLE
cannot add new osd

### ISSUE THREAD

**Issue by uhthomas** [rook/rook#12439]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

I'm trying to add a new OSD to the cluster, but rook doesn't seem to be able to.

```sh
❯ k -n rook-ceph logs rook-ceph-osd-prepare-talos-l94-p4c-x9htx --previou
...
[2023-06-28 10:37:01,999][ceph_volume.process][INFO  ] Running command: /usr/bin/ceph-bluestore-tool show-label --dev /dev/sdc
[2023-06-28 10:37:02,034][ceph_volume.process][INFO  ] stderr unable to read label for /dev/sdc: (5) Input/output error
[2023-06-28 10:37:02,035][ceph_volume.process][INFO  ] stderr 2023-06-28T10:37:02.028+0000 7f07a8040880 -1 bluestore(/dev/sdc) _read_bdev_label bad crc on label, expected 1633861745 != actual 3173833106
[2023-06-28 10:37:02,035][ceph_volume.process][INFO  ] Running command: /usr/sbin/blkid -c /dev/null -p /dev/sdc
[2023-06-28 10:37:02,051][ceph_volume.process][INFO  ] Running command: /usr/bin/ceph-authtool --gen-print-key
[2023-06-28 10:37:02,078][ceph_volume.process][INFO  ] Running command: /usr/bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring -i - osd new a5fbaf6b-cb39-4a2d-a6c3-c046898b091d
[2023-06-28 10:37:02,757][ceph_volume.process][INFO  ] stderr Error EEXIST: entity osd.1 exists but key does not match
[2023-06-28 10:37:02,757][ceph_volume.devices.raw.prepare][ERROR ] raw prepare was unable to complete
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 91, in safe_prepare
    self.prepare()
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 16, in is_root
    return func(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 125, in prepare
    osd_fsid, json.dumps(secrets))
  File "/usr/lib/python3.6/site-packages/ceph_volume/util/prepare.py", line 176, in create_id
    raise RuntimeError('Unable to create a new OSD id')
RuntimeError: Unable to create a new OSD id
[2023-06-28 10:37:02,759][ceph_volume.devices.raw.prepare][INFO  ] will rollback OSD ID creation
[2023-06-28 10:37:02,760][ceph_volume][ERROR ] exception caught by decorator
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 59, in newfunc
    return f(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 153, in main
    terminal.dispatch(self.mapper, subcommand_args)
  File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
    instance.main()
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/main.py", line 32, in main
    terminal.dispatch(self.mapper, self.argv)
  File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
    instance.main()
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 169, in main
    self.safe_prepare(self.args)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 91, in safe_prepare
    self.prepare()
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 16, in is_root
    return func(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 125, in prepare
    osd_fsid, json.dumps(secrets))
  File "/usr/lib/python3.6/site-packages/ceph_volume/util/prepare.py", line 176, in create_id
    raise RuntimeError('Unable to create a new OSD id')
RuntimeError: Unable to create a new OSD id
2023-06-28 10:37:05.198590 C | rookcmd: failed to configure devices: failed to initialize osd: failed to run ceph-volume raw command. Running command: /usr/bin/ceph-authtool --gen-print-key
Running command: /usr/bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring -i - osd new a5fbaf6b-cb39-4a2d-a6c3-c046898b091d
 stderr: Error EEXIST: entity osd.1 exists but key does not match
Traceback (most recent call last):
  File "/usr/sbin/ceph-volume", line 11, in <module>
    load_entry_point('ceph-volume==1.0.0', 'console_scripts', 'ceph-volume')()
  File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 41, in __init__
    self.main(self.argv)
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 59, in newfunc
    return f(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 153, in main
    terminal.dispatch(self.mapper, subcommand_args)
  File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
    instance.main()
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/main.py", line 32, in main
    terminal.dispatch(self.mapper, self.argv)
  File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
    instance.main()
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 169, in main
    self.safe_prepare(self.args)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 91, in safe_prepare
    self.prepare()
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 16, in is_root
    return func(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 125, in prepare
    osd_fsid, json.dumps(secrets))
  File "/usr/lib/python3.6/site-packages/ceph_volume/util/prepare.py", line 176, in create_id
    raise RuntimeError('Unable to create a new OSD id')
RuntimeError: Unable to create a new OSD id: exit status 1
```

**Expected behavior:**

rook should be able to add a new OSD to the cluster.

**How to reproduce it (minimal and precise):**

Not sure. I removed this device from one node (wiped, removed from the cluster) and put it in another. 

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary

**Logs to submit**:

* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

  To get logs, use `kubectl -n <namespace> logs <pod name>`
  When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
  Read [GitHub documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Cluster Status to submit**:

* Output of krew commands, if necessary

  To get the health of the cluster, use `kubectl rook-ceph health`
  To get the status of the cluster, use `kubectl rook-ceph ceph status`
  For more details, see the [Rook Krew Plugin](https://rook.io/docs/rook/latest-release/Troubleshooting/krew-plugin)

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Storage backend version (e.g. for ceph do `ceph -v`):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)):

**Comment by uhthomas** [comment]
`osd.1` does not exist, so I'm not sure what it's complaining about.

![image](https://github.com/rook/rook/assets/9749173/2004e6ef-e263-40e9-8004-b95c2d518389)

**Comment by gaord** [comment]
osd.1 looks like existing in monitor entity store, and it may be by previous deployment.
run "ceph auth ls" for a look.

**Comment by travisn** [comment]
@uhthomas You went through the [OSD Removal](https://rook.io/docs/rook/latest/Storage-Configuration/Advanced/ceph-osd-mgmt/#remove-an-osd) guide? This should have purged the OSD, but it seems the OSD auth wasn't fully purged.

**Comment by uhthomas** [comment]
I did follow the guide and made sure to purge the OSD and still saw this. I do see the the OSD exists in `ceph auth`.

```sh
bash-4.4$ ceph auth get osd.1
[osd.1]
        key = AQDNWJhkPP9jLBAAV6/pU0nxBSz4D0f98Wn5AQ==
        caps mgr = "allow profile osd"
        caps mon = "allow profile osd"
        caps osd = "allow *"
```

Though, there is no deployment for the osd and the osd does not show up anywhere else.

```sh
bash-4.4$ ceph osd ls
0
2
3
4
5
6
7
8
9
10
11
12
13
```

**Comment by sudharsanomprakash** [comment]
@uhthomas Did you try, ceph auth del osd.1 ?

**Comment by uhthomas** [comment]
> @uhthomas Did you try, ceph auth del osd.1 ?

This worked great, thank you. How did the cluster get into this state though? I followed the removal steps and clearly the OSD _was_ removed and purged given it no longer showed up on the crush map. So how come the auth for it still existed?

**Comment by sudharsanomprakash** [comment]
> 

It depends on how the osd was initially removed. If you set removeOSDsIfOutAndSafeToRemove=true in cluster.yaml , i guess this issue won't have occured in your scenario.

**Comment by uhthomas** [comment]
> removeOSDsIfOutAndSafeToRemove

I do have this.

https://github.com/uhthomas/automata/blob/b4c3cde4abe9eb6077dd6d9ed8936282a361087e/k8s/unwind/rook_ceph/ceph_cluster_list.cue#L80C1-L80C1

**Comment by uhthomas** [comment]
fwiw I just tried again and found the exact same thing. Here's what I did: 

1. Scale the number of replicas for the OSD deployment to 0.
2. Mark the OSD as out in Ceph.
3. `k rook-ceph rook purge-osd 3 --force`.
4. Clean the drive with `wipefs -a`.

I then verified the OSD was completely gone and even checked if the auth for it still existed, it didn't.

5. I physically moved the device from one node to another.
6. I updated the CephCluster object to include the device.

At this point I could see a new pod for the OSD was brought up, but was crashing. I checked Ceph and once again the auth for the OSD existed. I deleted it, the OSD container restarted on its own and resumed normal operation.

What's going on here?

**Comment by parth-gr** [comment]
Yaa that might be part of reconcile,
Is the OSDs fine for you or if not share the operator logs

**Comment by sudharsanomprakash** [comment]
> fwiw I just tried again and found the exact same thing. Here's what I did:
> 
> 1. Scale the number of replicas for the OSD deployment to 0.
> 2. Mark the OSD as out in Ceph.
> 3. `k rook-ceph rook purge-osd 3 --force`.
> 4. Clean the drive with `wipefs -a`.
> 
> I then verified the OSD was completely gone and even checked if the auth for it still existed, it didn't.
> 
> 5. I physically moved the device from one node to another.
> 6. I updated the CephCluster object to include the device.
> 
> At this point I could see a new pod for the OSD was brought up, but was crashing. I checked Ceph and once again the auth for the OSD existed. I deleted it, the OSD container restarted on its own and resumed normal operation.
> 
> What's going on here?

It should actually work with wipefs -a, but are you able to reproduce with zapping the device as mentioned [here](https://rook.io/docs/rook/v1.11/Getting-Started/ceph-teardown/#zapping-devices ) 

Todo: skip step 4 and after step 5, perform above procedure.

**Comment by uhthomas** [comment]
Can we reopen this, or open a new issue for the ceph toolbox? I _always_ have to manually run `ceph auth del` when I remove an OSD.

**Comment by travisn** [comment]
Reopening... To summarize the issue, you verify that the auth is gone after the purge action, but then it re-appears and you have to remove it again manually? So the `kubectl rook-ceph purge-osd` did its job, but the question is why it is re-appearing. I wonder if the auth is being found on the disk still. Is you wiping the disk before moving it to the new node?

**Comment by uhthomas** [comment]
@travisn I will have to verify the exact process, but please observe the current state of this cluster.

Some OSDs have been removed, namely osd.2.

![image](https://github.com/rook/rook/assets/9749173/ff4bbf60-5337-4f3a-b780-79d73b158f60)

Yet, the auth still remains.

![image](https://github.com/rook/rook/assets/9749173/8b2bd86b-9109-4271-bf51-8dc43beb6018)

This was also the case for osd.1, where I had to manually delete the auth entry.

It really makes it upset, and it will not provision the OSD until the auth entry is removed.

![image](https://github.com/rook/rook/assets/9749173/512b5194-0de8-44d1-bad6-fdfe733ddc1e)

**Comment by travisn** [comment]
@subhamkrai Can you try repro'ing this?

**Comment by subhamkrai** [comment]
> @subhamkrai Can you try repro'ing this?

I'll try to repro this sometime next week, currently busy with some other items.

**Comment by parth-gr** [comment]
@subhamkrai ^^
